### DISCUSSION TITLE
Are mandatory field part of what we call partial data?

### DISCUSSION THREAD

**Discussion by frederikhors** [urql-graphql/urql#2429] [Q&A] [answered]
I'm a bit confused as to how urql works with the required fields.

If I read how Graphql means that on it's docs page (https://graphql.org/learn/schema/#lists-and-non-null and https://graphql.org/learn/schema/#object-types-and-fields) they say:

> `String!` means that the field is _non-nullable_, meaning that the GraphQL service promises to always give you a value when you query this field. In the type language, we'll represent those with an exclamation mark.

and

> Here, we're using a `String` type and marking it as _Non-Null_ by adding an exclamation mark, `!` after the type name. This means that our server always expects to return a non-null value for this field, and if it ends up getting a null value that will actually trigger a GraphQL execution error, letting the client know that something has gone wrong.

Using urql many times I have fragments (used alone in queries) which do not have "required/mandatory" fields in it.

And even if I setup everything correctly using Schema Awareness and all that my "list > detail" behaviour does not work as expected: I expect to see data I already have (from list) in detail page but this does not happen because even if there is ONE missing required (in schema) field urql does not show alreay cached data at all and wait for backend call to finish.

I think urql should better use the cached data because even if that field is "mandatory" it doesn't matter: I will update that field when backend will answer to me but now I can show the user what I already have from the list.

Am I totally wrong?

Are you talking about this in https://github.com/FormidableLabs/urql/pull/444, right?

### EXAMPLE:

- schema.graphql:

```graphql
type Player {
  id: ID!
  createdAt: Time!
  name: String!
}
```

- player_queries.graphql:

```graphql
fragment PLAYER on Player {
  id
  createdAt
  name
}

query playerList($pagination: Pagination!) {
	playerList(pagination: $pagination) {
		nodes {
			...PLAYER
		}
	}
}

query playerByID($id: ID!) {
	playerByID(id: $id) {
		...PLAYER
	}
}
```

In this case if I remove `createdAt` from `PLAYER` fragment urql does not show me the detail I already have from the list!

Is this correct? Can we avoid this?

I'm forced to change my graphql schema removing exclamation points almost everywhere, where I need to have this functioning with urql.

**Comment by JoviDeCroock** [comment, marked as answer]
We have indeed gone over this quite a few times already. One missing field makes your UI indeterministic as you'd have to account for all permutations of missing data while the nullability and by extension types make you account for the absence of data.

So any non-nullable missing field missing implies one of these unexpected permutations for which we can't assure the UI taking it in account.

There are proposals to give the user more control like the current one in PR but making a distinction between just returning or being deterministic I will always opt for the latter #2018

So in practice:

when you code-gen an entity your typing will reflect what _is_ there when it resolves and nullables _can_ be missing or come later which is what we have optimized for, else your application  can become really complex depending on where you are navigating from

**Reply by frederikhors** [reply]
Thank you for answering me.

I do not understand why it'll be so complex. Is there an example of what you call "permutation"?

> we can't assure the UI taking it in account

I think this should not be a urql problem.

Can we simplify the solution like "if there is data already cached I'll show that when using `cache-and-network` regardless of whether or not it is mandatory on the schema"?

And anyway until you decide what to introduce for this "issue" **are you suggesting that I need to change the scheme to solve this problem now**?

**Reply by JoviDeCroock** [reply]
A permutation of data in this case is just a different set of missing fields, i.e. you go from `/profile` to `/user` as opposed to `/users` to `/user` in GraphQL we query a field in a lot of ways with different selection-sets so the amount of missing fields (and the fields themselves) can be different every time.

I would argue that it _is_ a problem urql should tackle as else it would be a very common issue-topic where people ask why they need to guard for so many properties while they expect them to be there, all of a sudden you would have to explain that they need to account for navigating from any selection set (for instance getting just an amount of items in cart to the cart view) and that that's expected 😅 that does not feel like a good developer experience at all

**Reply by frederikhors** [reply]
Ok. Can we add a info in console in development mode to warn the developer that some fragment miss some "mandatory" fields?
